frontend/src/features/verifiable-fields/VerifiableFieldProvider.tsx

```diff
@@ -0,0 +1,139 @@
+import { useCallback, useState } from 'react'
+import {
+  ControllerRenderProps,
+  useFormContext,
+  useWatch,
+} from 'react-hook-form'
+
+import { FormFieldWithId } from '~shared/types/field'
+import { isMobilePhoneNumber } from '~shared/utils/phone-num-validation'
+
+import { BaseFieldProps } from '~templates/Field/FieldContainer'
+
+import {
+  VerifiableFieldBase,
+  VerifiableFieldInput,
+  VerifiableFieldSchema,
+} from './types'
+import { VerifiableFieldContext } from './VerifiableFieldContext'
+
+export interface VerifiableFieldProps extends BaseFieldProps {
+  schema: VerifiableFieldSchema<FormFieldWithId<VerifiableFieldBase>>
+}
+
+export interface VerifiableFieldProviderProps extends VerifiableFieldProps {
+  children: React.ReactNode
+}
+
+export const VerifiableFieldProvider = ({
+  schema,
+  children,
+}: VerifiableFieldProviderProps): JSX.Element => {
+  const [isVfnBoxOpen, setIsVfnBoxOpen] = useState(false)
+
+  const { control, setError, getValues, setValue, setFocus, clearErrors } =
+    useFormContext()
+  const currentSignature: VerifiableFieldInput = useWatch({
+    name: `${schema._id}.signature`,
+    control,
+  })
+
+  /**
+   * Keeps track of already completed numbers to signatures
+   */
+  const [mapNumberToSignature, setMapNumberToSignature] = useState<
+    Record<string, string>
+  >({})
+
+  const handleInputChange = useCallback(
+    (onChange: ControllerRenderProps['onChange']) => (value?: string) => {
+      // Prevent action when multiple onChange is called with the same value
+      // This can happen when input is blurred, since onChange is also called by
+      // react-hook-form when that happen.
+      if (getValues(schema._id)?.value === value) return
+      if (isVfnBoxOpen) {
+        setIsVfnBoxOpen(false)
```

Comment:
can we add a comment here for future devs pls? saying that if the input changes then the user needs to request re-verification, unless the current input is already verified (which is dealt with below)

Reply:
Done